### PR TITLE
add tRPC

### PR DESCRIPTION
**Pull request by KATT**
This PR adds tRPC which automatically infers the types from your API-function into the frontend.

## As a reference I rewrote `/bookings` + `/availability`-pages

- `/bookings`-page and `Shell` to use `trpc.useQuery()` rather than custom fetch hook and `getServerSideProps`
- The way to rewrite is basically to cut+pasted `getServerSideProps` into a function in `routes`

### Benefits

- Makes first load of "skeleton" edge-cachable so it feels more snappy
- When navigating between pages it won't require a slow full refetch
- Will be less flaky and fetch errors are automatically retried
- Automatic type safety all the way from DB to frontend when doing react-query calls
- "Refetch on focus" - so when re-opening the page it will automatically refresh all underlying data

## Extra fixes on this branch

- Fixed `callbackUrl` on login so it redirects to the right page after signing in

## How to test it

- Checkout this branch
- Run `yarn dx` to start 

## Next steps

#615 depends on this branch (will make everything lightning fast ⚡️)

## More info 

for more info on tRPC have a look at https://trpc.io or for a deep-dive look at https://www.youtube.com/watch?v=-_GZJ3xwYOw

feel free to ping me on Slack - maybe I should make a loom?

### CODE REVIEW

File: pages/[user].tsx [hunk 1]
```diff
@@ -4,14 +4,18 @@ import Theme from "@components/Theme";
...
 import Link from "next/link";
 import React from "react";
 
 export default function User(props: inferSSRProps<typeof getServerSideProps>) {
   const { isReady } = Theme(props.user.theme);
-
+  // FIXME delete me before merge
+  const query1 = trpc.useQuery(["test.hello"]);
+  const query2 = trpc.useQuery(["test.hello", "alex"]);
+  console.log("result", { query1, query2 });
```
Comment (KATT) [inline pages/[user].tsx:18]: This is just for show
```suggestion
```

**Comment by KATT** [comment]
_Converting to draft as I'm dealing with merge conflicts_

**Review by pumfleet — APPROVED** [review]
Haven't tested this locally but the code looks great :)

File: server/routers/viewer.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        },
+        orderBy: {
+          startTime: "asc",
+        },
+      });
+
+      const bookings = bookingsQuery.reverse().map((booking) => {
+        return {
+          ...booking,
+          startTime: booking.startTime.toISOString(),
```
Comment (emrysal) [inline server/routers/viewer.tsx:74]: Is this necessary (tRPC has issues dealing with Date/Dayjs objects?)

File: server/routers/viewer.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    return next({
+      ctx: {
+        ...ctx,
+        // session value is known to be non-null now
+        user,
+      },
+    });
+  })
+  .query("me", {
```
Comment (emrysal) [inline server/routers/viewer.tsx:22]: How should we go about splitting these into files?

**Review by emrysal — APPROVED** [review]

File: server/routers/viewer.tsx (hunk 2 above)
Comment (KATT) [reply server/routers/viewer.tsx:74]: I just copy-pasted what was already in `getServerSideProps`.

In tRPC we could actually just use a `Date`-object if we add superjson.

Ref: https://trpc.io/docs/data-transformers

File: server/routers/viewer.tsx (hunk 3 above)
Comment (KATT) [reply server/routers/viewer.tsx:22]: The same way we do on the app/root router to separate `viewer` in it's own file

https://github.com/calendso/calendso/blob/fb10bea04a0747e8abe005591b2780939c2da676/server/routers/_app.ts#L13-L24

**One can either do** 

```ts
export const myRouter = createRouter()
  .merge("sub.", subRouter); // <-- prefixes all procedures with `sub.`
```
**Or**

```ts
export const myRouter = createRouter()
  .merge(subRouter); // <-- does not prefix any procedures
```

And this works on the `viewerRouter` too, so if that file got insanely big or if we wanted to do distinct parts separately, they could be handled in their own file

**Review by PeerRich — APPROVED** [review]

**Pull request merged**